backend/dal/Services/Admin/Concrete/AgencyService.cs

```diff
@@ -100,35 +100,133 @@ public Agency Get(int id)
             return this.Context.Agencies.Find(id) ?? throw new KeyNotFoundException();
         }
 
+        /// <summary>
+        /// Add a new agency to the datasource.
+        /// </summary>
+        /// <param name="agency"></param>
+        public override void Add(Agency agency)
+        {
+            agency.ThrowIfNull(nameof(agency));
+
+            base.Add(agency);
+            this.Context.Entry(agency).State = EntityState.Detached;
+
+            // If the agency has been added as a sub-agency, then all users who are associated with the parent agency need to be updated.
+            if (agency.ParentId.HasValue)
+            {
+                var users = this.Context.Users.AsNoTracking().Where(u => u.Agencies.Any(a => a.AgencyId == agency.ParentId)).ToArray();
+                users.ForEach(u => agency.Users.Add(new UserAgency(u, agency)));
+            }
+        }
+
         /// <summary>
         /// Updates the specified agency in the datasource.
         /// </summary>
-        /// <param name="entity"></param>
-        /// <exception type="KeyNotFoundException">Entity does not exist in the datasource.</exception>
+        /// <param name="agency"></param>
+        /// <exception type="KeyNotFoundException">agency does not exist in the datasource.</exception>
         /// <returns></returns>
-        public override void Update(Agency entity)
+        public override void Update(Agency agency)
         {
-            entity.ThrowIfNull(nameof(entity));
+            agency.ThrowIfNull(nameof(agency));
+
+            var original = this.Context.Agencies.Find(agency.Id) ?? throw new KeyNotFoundException();
+            var updatedUsers = new List<User>();
+
+            // If the agency has become a sub-agency, or a parent-agency then users will need to be updated.
+            if (original.ParentId.HasValue && !agency.ParentId.HasValue)
+            {
+                // This agency has become a parent agency, all users associated with it will be removed
+                // because users should only be associated with a single parent.
+                updatedUsers.AddRange(this.Context.Users.Where(u => u.Agencies.Any(a => a.AgencyId == original.ParentId)));
+
+                // Remove any user who are linked to the agency.
+                var users = this.Context.Users.Include(u => u.Agencies).Where(u => u.Agencies.Any(ua => ua.AgencyId == agency.Id));
+                users.ForEach(u => u.Agencies.Clear());
+                this.Context.UpdateRange(users);
+            }
+            else if (!original.ParentId.HasValue && agency.ParentId.HasValue)
+            {
+                // This agency has become a sub-agency, all original users need their list of agencies reduced to only this agency.
```

Comment:
only to this agency -> So should this logic be clearing the unrelated agencies then?